analytics/transform.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return nil, err
+	}
+
+	jsonified, err := json.Marshal(mapified)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error marshaling to JSON")
+	}
+	return jsonified, nil
+}
+
+// ToJsonWithGeo adds the geo_location field, and transforms a valid tsv string Snowplow event to a JSON object.
+func ToJsonWithGeo(event string) ([]byte, error) {
+	mapified, err := ToMapWithGeo(event)
+	if err != nil {
+		return nil, err
+	}
+
+	jsonified, err := json.Marshal(mapified)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error marshaling to JSON")
+	}
+	return jsonified, nil
+}
+
+// Design question: If the field(s) provided to the below Get functions are (all) empty - should it return an error, a zero value, or nil, nil?
+// Currently it returns nil, and an error for GetValue, and an empty map/json with no error for GetSubsetMap and GetSubsetJSON...
+// I have a natrual aversion to zero value but not sure if it's well founded.
+
+// Design decision made: unstruct_event, contexts and derived_contexts return the structure `{"unstruct_event_com_acme_event_1": {"field1": "value1"}}`
+
+// GetValue returns the value for a provided atomic field, without processing the rest of the event.
+// For unstruct_event, it returns a map of only the data for the unstruct event.
+func GetValue(event string, field string) (interface{}, error) {
+
+	// TODO: DRY HERE
+	record := strings.Split(event, "\t")
+	if len(record) != 131 { // leave hardcoded or not?
+		return nil, errors.New("Cannot get value from event - wrong number of fields")
+	} else {
+		index, ok := indexMap[field]
+		if !ok {
+			return nil, errors.New(fmt.Sprintf("Key %s not a valid atomic field", field))
+		}
+		if record[index] == "" {
+			return nil, errors.New(fmt.Sprintf("Field %s is empty", field))
+		}
+		kvPairs, err := enrichedEventFieldTypes[index].ParseFunction(enrichedEventFieldTypes[index].Key, record[index])
+		if err != nil {
+			return nil, err
+		}
+		if field == "contexts" || field == "derived_contexts" || field == "unstruct_event" {
+			// TODO: DRY HERE TOO?
+			output := make(map[string]interface{})
+			for _, pair := range kvPairs {
+				output[pair.Key] = pair.Value
+			}
+			return output, nil
+		}
+		return kvPairs[0].Value, nil
+	}
+}
+
+// GetSubsetMap returns a map of a subset of the event, containing only the atomic fields provided, without processing the rest of the event.
+// For custom events and contexts, only "unstruct_event", "contexts", or "derived_contexts" may be provided, which will produce the entire data object for that field.
+// For contexts, the resultant map will contain all occurrences of all contexts within the provided field.
+func GetSubsetMap(event string, fields []string) (map[string]interface{}, error) {
+	// TODO: Same error handling issue as above - what should the behaviour be in case of no value?
+
+	// TODO: DRY HERE
+	record := strings.Split(event, "\t")
+	if len(record) != 131 { // leave hardcoded or not?
```

Comment:
I'd usually pull "magic numbers" into a constant somewhere.

Reply:
So the other sdks would use len(enrichedEventFieldTypes). I don't think this is a terribly consequential decision so sticking it into a const and moving on. :)